kong/pdk/vault.lua

```diff
@@ -0,0 +1,404 @@
+-- ```
+--
+-- This function will return following table:
+--
+-- ```lua
+-- {
+--   name     = "env",  -- name of the Vault entity or Vault strategy
+--   resource = "cert", -- resource where secret is stored
+--   key      = "key",  -- key to lookup if the resource is secret object
+--   config   = {       -- if there are any config options specified
+--     prefix = "SSL_"
+--   },
+--   version  = 1       -- if the version is specified
+-- }
+-- ```
+--
+-- @function kong.vault.parse_reference
+-- @tparam   string      reference  reference to parse
+-- @treturn  table|nil              a table containing each component of the reference, or `nil` on error
+-- @treturn  string|nil             error message on failure, otherwise `nil`
+--
+-- @usage
+-- local ref, err = kong.vault.parse_reference("{vault://env/cert/key?prefix=SSL_#1}") -- table
+local function parse_reference(reference)
+  if not is_reference(reference) then
+    return nil, fmt("not a reference [%s]", tostring(reference))
+  end
+
+  local url, err = parse_url(sub(reference, 2, -2))
+  if not url then
+    return nil, fmt("reference is not url (%s) [%s]", err, reference)
+  end
+
+  local name = url.host
+  if not name then
+    return nil, fmt("reference url is missing host [%s]", reference)
+  end
+
+  local path = url.path
+  if not path then
+    return nil, fmt("reference url is missing path [%s]", reference)
+  end
+
+  local resource = sub(path, 2)
+  if resource == "" then
+    return nil, fmt("reference url has empty path [%s]", reference)
+  end
+
+  local version = url.fragment
+  if version then
+    version = tonumber(version, 10)
+    if not version then
+      return nil, fmt("reference url has invalid version [%s]", reference)
+    end
+  end
+
+  local key
+  local parts = parse_path(resource)
+  local count = #parts
+  if count == 1 then
+    resource = unescape_uri(parts[1])
+
+  else
+    resource = unescape_uri(concat(parts, "/", 1, count - 1))
+    if parts[count] ~= "" then
+      key = unescape_uri(parts[count])
+    end
+  end
+
+  if resource == "" then
+    return nil, fmt("reference url has invalid path [%s]", reference)
+  end
+
+  local config
+  local query = url.query
+  if query and query ~= "" then
+    config = decode_args(query)
+  end
+
+  return {
+    name = url.host,
+    resource = resource,
+    key = key,
+    config = config,
+    version = version,
+  }
+end
+
+
+---
+-- Resolves the passed in reference and returns the value of it.
+--
+-- @function kong.vault.get
+-- @tparam   string      reference  reference to resolve
+-- @treturn  string|nil             resolved value of the reference
+-- @treturn  string|nil             error message on failure, otherwise `nil`
+--
+-- @usage
+-- local value, err = kong.vault.get("{vault://env/cert/key}")
+local function get(reference)
+  local opts, err = parse_reference(reference)
+  if err then
+    return nil, err
+  end
+
+  local value = LRU:get(reference)
```

Comment:
If we receive a cached value, we can immediately return it
```suggestion
  local value = LRU:get(reference)
  if value then
    return value
  end
```